### PR TITLE
Refactor story_store

### PR DESCRIPTION
**Pull request by shilman**
Issue: #7101 

## What I did

Refactor `story_store` in prep for docs mode:

- don't touch URL
- only emit events, don't listen on them
- add tests

All of these changes are positive independent of docs mode.

### CODE REVIEW

File: lib/client-api/src/story_store.js [hunk 1]
```diff
@@ -126,8 +95,8 @@ export default class StoryStore extends EventEmitter {
     );
   }
 
-  setSelection = data => {
-    this._selection = data;
+  setSelection = ({ storyId }) => {
+    this._selection = { storyId };
```
Comment (ndelangen) [inline lib/client-api/src/story_store.js]: If the only datapoint is the storyId, Why make it an object?

File: lib/client-api/src/story_store.js [hunk 2]
```diff
@@ -126,8 +95,8 @@ export default class StoryStore extends EventEmitter {
...
     setTimeout(() => this.emit(Events.STORY_RENDER), 1);
```
Comment (ndelangen) [inline lib/client-api/src/story_store.js]: If you're going to refactor this, I think we could remove the inheritance from EventEmitter and use the channel instead?

Just to have less moving parts and things to subscribe to?

File: lib/core/src/client/preview/start.js [hunk 3]
```diff
@@ -187,10 +194,8 @@ export default function start(render, { decorateStory } = {}) {
         storyId = deprecatedToId(kind, name);
       }
 
-      const data = storyStore.fromId(storyId);
-
-      storyStore.setSelection(data);
-      storyStore.setPath(storyId);
+      storyStore.setSelection({ storyId });
+      setPath({ storyId });
```
Comment (ndelangen) [inline lib/core/src/client/preview/start.js]: Nice! i like how storyStore is no longer responsible for path management

File: lib/core/src/client/preview/start.js [hunk 4]
```diff
@@ -205,6 +210,11 @@ export default function start(render, { decorateStory } = {}) {
     };
   }
 
+  storyStore.on(Events.STORY_INIT, () => {
```
Comment (ndelangen) [inline lib/core/src/client/preview/start.js]: Could be on the channel instead?

File: lib/core/src/client/preview/url.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { history, document } from 'global';
+import qs from 'qs';
+import { toId } from '@storybook/router/utils';
+
+export function pathToId(path) {
```
Comment (ndelangen) [inline lib/core/src/client/preview/url.js]: Could this be part of `lib/router/utils` ?

**Review by ndelangen — CHANGES_REQUESTED** [review]

File: lib/client-api/src/story_store.js (hunk 1 above)
Comment (tmeasday) [reply lib/client-api/src/story_store.js]: I think @shilman has secret plans and clever tricks he is not telling you about in this PR

Comment (tmeasday) [reply lib/client-api/src/story_store.js]: (The answer is later there'll be `docs: boolean` as a second datapoint)

File: lib/client-api/src/story_store.js (hunk 2 above)
Comment (tmeasday) [reply lib/client-api/src/story_store.js]: Probably yes. We weren’t sure of the full implications (eg in jest where there is no channel) so we held off doing that

**Comment by ndelangen** [comment]
@shilman want me to pick this up?

I think more merge-conflicts will be happening if you let this collect dust some more.

**Comment by shilman** [comment]
@ndelangen No worries, I'll update it this weekend.

File: lib/client-api/src/story_store.js (hunk 2 above)
Comment (ndelangen) [reply lib/client-api/src/story_store.js]: We'll do this later

File: lib/core/src/client/preview/start.js (hunk 4 above)
Comment (ndelangen) [reply lib/core/src/client/preview/start.js]: we'll do this later

**Comment by ndelangen** [comment]
ping @shilman

**Comment by shilman** [comment]
Roger. Will update this with the changes we made together the other day. Also found a bug in our work & fixed in my branch 😉

**Comment by shilman** [comment]
@ndelangen @tmeasday @igor-dv First PR for https://github.com/storybookjs/storybook/issues/7101

File: lib/client-api/src/story_store.js [hunk 6]
```diff
@@ -1,17 +1,15 @@
...
-import qs from 'qs';
 import EventEmitter from 'eventemitter3';
 import memoize from 'memoizerific';
 import debounce from 'lodash/debounce';
 import { stripIndents } from 'common-tags';
 
 import Events from '@storybook/core-events';
 import { logger } from '@storybook/client-logger';
-import { toId } from '@storybook/router/utils';
+// import { toId } from '@storybook/router/utils';
```
Comment (ndelangen) [inline lib/client-api/src/story_store.js]: Some cleanup and then this looks good to go to me

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
@Jessica-Koch heads-up, some changes incoming on client-api (which you're currently migrating to typescript).

**Pull request merged**